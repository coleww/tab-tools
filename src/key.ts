import { ALL_NOTES, KEY_INTERVALS, TabData } from './constants';
import { getNote } from './note';

// 'a min': ['a', 'b', 'c', 'd', 'e', 'f', 'g']
function buildKeyMap(): Record<string, string[]> {
  return ALL_NOTES.reduce<Record<string, string[]>>(
    (keyMap, rootNote) => {
      Object.entries(KEY_INTERVALS).forEach(([keyType, intervals]) => {
        const keyName = `${rootNote} ${keyType}`;
        const keyNotes = intervals.reduce(
          (acc, interval) => {
            acc.push(getNote(rootNote, `${interval}`));
            return acc;
          },
          [rootNote]
        );

        keyMap[keyName] = keyNotes;
      });
      return keyMap;
    },
    { chromatic: ALL_NOTES }
  );
}

export const keyMap = buildKeyMap();

export function getPossibleKeys(tabData: TabData) {
  const tabNotes = getUniqueNotes(tabData);
  return Object.keys(keyMap).filter(key => {
    return tabNotes.every(note => keyMap[key]?.includes(note));
  });
}

export function getUniqueNotes(tabData: TabData): string[] {
  return tabData.data
    .reduce<string[]>((bigAcc, string, i) => {
      const rootNote = tabData.tuning[i];
      return bigAcc.concat(
        string.reduce<string[]>((lilAcc, fret) => {
          if (fret && rootNote) {
            const note = getNote(rootNote, fret);
            if (note) {
              lilAcc.push(note);
            }
          }

          return lilAcc;
        }, [])
      );
    }, [])
    .filter((note, i, arr) => arr.indexOf(note) === i)
    .sort();
}
