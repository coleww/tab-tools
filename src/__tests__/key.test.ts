import { mockTabData } from './fixtures';
import { getUniqueNotes, getPossibleKeys } from '../key';
import { KEY_MAP } from '../key';
describe('getUniqueNotes', () => {
  it('returns unique notes for tab data', () => {
    expect(getUniqueNotes(mockTabData)).toStrictEqual([
      'a',
      'b',
      'c',
      'e',
      'f',
      'g',
    ]);
  });
});

describe('keyMap', () => {
  it('maps keys to note values', () => {
    expect(KEY_MAP['a min']).toStrictEqual(['a', 'b', 'c', 'd', 'e', 'f', 'g']);
  });

  it('matches snapshot', () => {
    expect(KEY_MAP).toMatchSnapshot();
  });
});

describe('getPossibleKeys', () => {
  it('gets keys that contain all the notes in the tab data', () => {
    expect(getPossibleKeys(mockTabData)).toStrictEqual([
      'chromatic',
      'a min',
      'c maj',
    ]);
  });

  it('handles variadic input', () => {
    expect(
      getPossibleKeys(
        {
          ...mockTabData,
          tuning: ['a', 'c', 'c', 'b'],
        },
        {
          ...mockTabData,
          tuning: ['g', 'c', 'c', 'b'],
        }
      )
    ).toStrictEqual(['chromatic', 'c melMin']);
  });
});
