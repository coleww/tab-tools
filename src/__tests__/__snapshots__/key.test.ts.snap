// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`keyMap matches snapshot 1`] = `
{
  "a harmMin": [
    "a",
    "b",
    "c",
    "d",
    "e",
    "f",
    "g#",
  ],
  "a maj": [
    "a",
    "b",
    "c#",
    "d",
    "e",
    "f#",
    "g#",
  ],
  "a melMin": [
    "a",
    "b",
    "c",
    "d",
    "e",
    "f#",
    "g#",
  ],
  "a min": [
    "a",
    "b",
    "c",
    "d",
    "e",
    "f",
    "g",
  ],
  "a oct1": [
    "a",
    "a#",
    "c",
    "c#",
    "d#",
    "e",
    "f#",
  ],
  "a oct2": [
    "a",
    "b",
    "c",
    "d",
    "d#",
    "f",
    "f#",
    "g#",
  ],
  "a pentMaj": [
    "a",
    "b",
    "c#",
    "e",
    "f#",
  ],
  "a pentMin": [
    "a",
    "c",
    "d",
    "e",
    "g",
  ],
  "a wholeTone": [
    "a",
    "b",
    "c#",
    "d#",
    "f",
    "g",
  ],
  "a# harmMin": [
    "a#",
    "c",
    "c#",
    "d#",
    "f",
    "f#",
    "a",
  ],
  "a# maj": [
    "a#",
    "c",
    "d",
    "d#",
    "f",
    "g",
    "a",
  ],
  "a# melMin": [
    "a#",
    "c",
    "c#",
    "d#",
    "f",
    "g",
    "a",
  ],
  "a# min": [
    "a#",
    "c",
    "c#",
    "d#",
    "f",
    "f#",
    "g#",
  ],
  "a# oct1": [
    "a#",
    "b",
    "c#",
    "d",
    "e",
    "f",
    "g",
  ],
  "a# oct2": [
    "a#",
    "c",
    "c#",
    "d#",
    "e",
    "f#",
    "g",
    "a",
  ],
  "a# pentMaj": [
    "a#",
    "c",
    "d",
    "f",
    "g",
  ],
  "a# pentMin": [
    "a#",
    "c#",
    "d#",
    "f",
    "g#",
  ],
  "a# wholeTone": [
    "a#",
    "c",
    "d",
    "e",
    "f#",
    "g#",
  ],
  "b harmMin": [
    "b",
    "c#",
    "d",
    "e",
    "f#",
    "g",
    "a#",
  ],
  "b maj": [
    "b",
    "c#",
    "d#",
    "e",
    "f#",
    "g#",
    "a#",
  ],
  "b melMin": [
    "b",
    "c#",
    "d",
    "e",
    "f#",
    "g#",
    "a#",
  ],
  "b min": [
    "b",
    "c#",
    "d",
    "e",
    "f#",
    "g",
    "a",
  ],
  "b oct1": [
    "b",
    "c",
    "d",
    "d#",
    "f",
    "f#",
    "g#",
  ],
  "b oct2": [
    "b",
    "c#",
    "d",
    "e",
    "f",
    "g",
    "g#",
    "a#",
  ],
  "b pentMaj": [
    "b",
    "c#",
    "d#",
    "f#",
    "g#",
  ],
  "b pentMin": [
    "b",
    "d",
    "e",
    "f#",
    "a",
  ],
  "b wholeTone": [
    "b",
    "c#",
    "d#",
    "f",
    "g",
    "a",
  ],
  "c harmMin": [
    "c",
    "d",
    "d#",
    "f",
    "g",
    "g#",
    "b",
  ],
  "c maj": [
    "c",
    "d",
    "e",
    "f",
    "g",
    "a",
    "b",
  ],
  "c melMin": [
    "c",
    "d",
    "d#",
    "f",
    "g",
    "a",
    "b",
  ],
  "c min": [
    "c",
    "d",
    "d#",
    "f",
    "g",
    "g#",
    "a#",
  ],
  "c oct1": [
    "c",
    "c#",
    "d#",
    "e",
    "f#",
    "g",
    "a",
  ],
  "c oct2": [
    "c",
    "d",
    "d#",
    "f",
    "f#",
    "g#",
    "a",
    "b",
  ],
  "c pentMaj": [
    "c",
    "d",
    "e",
    "g",
    "a",
  ],
  "c pentMin": [
    "c",
    "d#",
    "f",
    "g",
    "a#",
  ],
  "c wholeTone": [
    "c",
    "d",
    "e",
    "f#",
    "g#",
    "a#",
  ],
  "c# harmMin": [
    "c#",
    "d#",
    "e",
    "f#",
    "g#",
    "a",
    "c",
  ],
  "c# maj": [
    "c#",
    "d#",
    "f",
    "f#",
    "g#",
    "a#",
    "c",
  ],
  "c# melMin": [
    "c#",
    "d#",
    "e",
    "f#",
    "g#",
    "a#",
    "c",
  ],
  "c# min": [
    "c#",
    "d#",
    "e",
    "f#",
    "g#",
    "a",
    "b",
  ],
  "c# oct1": [
    "c#",
    "d",
    "e",
    "f",
    "g",
    "g#",
    "a#",
  ],
  "c# oct2": [
    "c#",
    "d#",
    "e",
    "f#",
    "g",
    "a",
    "a#",
    "c",
  ],
  "c# pentMaj": [
    "c#",
    "d#",
    "f",
    "g#",
    "a#",
  ],
  "c# pentMin": [
    "c#",
    "e",
    "f#",
    "g#",
    "b",
  ],
  "c# wholeTone": [
    "c#",
    "d#",
    "f",
    "g",
    "a",
    "b",
  ],
  "chromatic": [
    "a",
    "a#",
    "b",
    "c",
    "c#",
    "d",
    "d#",
    "e",
    "f",
    "f#",
    "g",
    "g#",
  ],
  "d harmMin": [
    "d",
    "e",
    "f",
    "g",
    "a",
    "a#",
    "c#",
  ],
  "d maj": [
    "d",
    "e",
    "f#",
    "g",
    "a",
    "b",
    "c#",
  ],
  "d melMin": [
    "d",
    "e",
    "f",
    "g",
    "a",
    "b",
    "c#",
  ],
  "d min": [
    "d",
    "e",
    "f",
    "g",
    "a",
    "a#",
    "c",
  ],
  "d oct1": [
    "d",
    "d#",
    "f",
    "f#",
    "g#",
    "a",
    "b",
  ],
  "d oct2": [
    "d",
    "e",
    "f",
    "g",
    "g#",
    "a#",
    "b",
    "c#",
  ],
  "d pentMaj": [
    "d",
    "e",
    "f#",
    "a",
    "b",
  ],
  "d pentMin": [
    "d",
    "f",
    "g",
    "a",
    "c",
  ],
  "d wholeTone": [
    "d",
    "e",
    "f#",
    "g#",
    "a#",
    "c",
  ],
  "d# harmMin": [
    "d#",
    "f",
    "f#",
    "g#",
    "a#",
    "b",
    "d",
  ],
  "d# maj": [
    "d#",
    "f",
    "g",
    "g#",
    "a#",
    "c",
    "d",
  ],
  "d# melMin": [
    "d#",
    "f",
    "f#",
    "g#",
    "a#",
    "c",
    "d",
  ],
  "d# min": [
    "d#",
    "f",
    "f#",
    "g#",
    "a#",
    "b",
    "c#",
  ],
  "d# oct1": [
    "d#",
    "e",
    "f#",
    "g",
    "a",
    "a#",
    "c",
  ],
  "d# oct2": [
    "d#",
    "f",
    "f#",
    "g#",
    "a",
    "b",
    "c",
    "d",
  ],
  "d# pentMaj": [
    "d#",
    "f",
    "g",
    "a#",
    "c",
  ],
  "d# pentMin": [
    "d#",
    "f#",
    "g#",
    "a#",
    "c#",
  ],
  "d# wholeTone": [
    "d#",
    "f",
    "g",
    "a",
    "b",
    "c#",
  ],
  "e harmMin": [
    "e",
    "f#",
    "g",
    "a",
    "b",
    "c",
    "d#",
  ],
  "e maj": [
    "e",
    "f#",
    "g#",
    "a",
    "b",
    "c#",
    "d#",
  ],
  "e melMin": [
    "e",
    "f#",
    "g",
    "a",
    "b",
    "c#",
    "d#",
  ],
  "e min": [
    "e",
    "f#",
    "g",
    "a",
    "b",
    "c",
    "d",
  ],
  "e oct1": [
    "e",
    "f",
    "g",
    "g#",
    "a#",
    "b",
    "c#",
  ],
  "e oct2": [
    "e",
    "f#",
    "g",
    "a",
    "a#",
    "c",
    "c#",
    "d#",
  ],
  "e pentMaj": [
    "e",
    "f#",
    "g#",
    "b",
    "c#",
  ],
  "e pentMin": [
    "e",
    "g",
    "a",
    "b",
    "d",
  ],
  "e wholeTone": [
    "e",
    "f#",
    "g#",
    "a#",
    "c",
    "d",
  ],
  "f harmMin": [
    "f",
    "g",
    "g#",
    "a#",
    "c",
    "c#",
    "e",
  ],
  "f maj": [
    "f",
    "g",
    "a",
    "a#",
    "c",
    "d",
    "e",
  ],
  "f melMin": [
    "f",
    "g",
    "g#",
    "a#",
    "c",
    "d",
    "e",
  ],
  "f min": [
    "f",
    "g",
    "g#",
    "a#",
    "c",
    "c#",
    "d#",
  ],
  "f oct1": [
    "f",
    "f#",
    "g#",
    "a",
    "b",
    "c",
    "d",
  ],
  "f oct2": [
    "f",
    "g",
    "g#",
    "a#",
    "b",
    "c#",
    "d",
    "e",
  ],
  "f pentMaj": [
    "f",
    "g",
    "a",
    "c",
    "d",
  ],
  "f pentMin": [
    "f",
    "g#",
    "a#",
    "c",
    "d#",
  ],
  "f wholeTone": [
    "f",
    "g",
    "a",
    "b",
    "c#",
    "d#",
  ],
  "f# harmMin": [
    "f#",
    "g#",
    "a",
    "b",
    "c#",
    "d",
    "f",
  ],
  "f# maj": [
    "f#",
    "g#",
    "a#",
    "b",
    "c#",
    "d#",
    "f",
  ],
  "f# melMin": [
    "f#",
    "g#",
    "a",
    "b",
    "c#",
    "d#",
    "f",
  ],
  "f# min": [
    "f#",
    "g#",
    "a",
    "b",
    "c#",
    "d",
    "e",
  ],
  "f# oct1": [
    "f#",
    "g",
    "a",
    "a#",
    "c",
    "c#",
    "d#",
  ],
  "f# oct2": [
    "f#",
    "g#",
    "a",
    "b",
    "c",
    "d",
    "d#",
    "f",
  ],
  "f# pentMaj": [
    "f#",
    "g#",
    "a#",
    "c#",
    "d#",
  ],
  "f# pentMin": [
    "f#",
    "a",
    "b",
    "c#",
    "e",
  ],
  "f# wholeTone": [
    "f#",
    "g#",
    "a#",
    "c",
    "d",
    "e",
  ],
  "g harmMin": [
    "g",
    "a",
    "a#",
    "c",
    "d",
    "d#",
    "f#",
  ],
  "g maj": [
    "g",
    "a",
    "b",
    "c",
    "d",
    "e",
    "f#",
  ],
  "g melMin": [
    "g",
    "a",
    "a#",
    "c",
    "d",
    "e",
    "f#",
  ],
  "g min": [
    "g",
    "a",
    "a#",
    "c",
    "d",
    "d#",
    "f",
  ],
  "g oct1": [
    "g",
    "g#",
    "a#",
    "b",
    "c#",
    "d",
    "e",
  ],
  "g oct2": [
    "g",
    "a",
    "a#",
    "c",
    "c#",
    "d#",
    "e",
    "f#",
  ],
  "g pentMaj": [
    "g",
    "a",
    "b",
    "d",
    "e",
  ],
  "g pentMin": [
    "g",
    "a#",
    "c",
    "d",
    "f",
  ],
  "g wholeTone": [
    "g",
    "a",
    "b",
    "c#",
    "d#",
    "f",
  ],
  "g# harmMin": [
    "g#",
    "a#",
    "b",
    "c#",
    "d#",
    "e",
    "g",
  ],
  "g# maj": [
    "g#",
    "a#",
    "c",
    "c#",
    "d#",
    "f",
    "g",
  ],
  "g# melMin": [
    "g#",
    "a#",
    "b",
    "c#",
    "d#",
    "f",
    "g",
  ],
  "g# min": [
    "g#",
    "a#",
    "b",
    "c#",
    "d#",
    "e",
    "f#",
  ],
  "g# oct1": [
    "g#",
    "a",
    "b",
    "c",
    "d",
    "d#",
    "f",
  ],
  "g# oct2": [
    "g#",
    "a#",
    "b",
    "c#",
    "d",
    "e",
    "f",
    "g",
  ],
  "g# pentMaj": [
    "g#",
    "a#",
    "c",
    "d#",
    "f",
  ],
  "g# pentMin": [
    "g#",
    "b",
    "c#",
    "d#",
    "f#",
  ],
  "g# wholeTone": [
    "g#",
    "a#",
    "c",
    "d",
    "e",
    "f#",
  ],
}
`;
